File: 056. 合并区间.py
import copy


class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        #先排序，然后建一个新列表
        # 无重叠->放入
        # 有重叠->修改上一个进入的元素的右边界
        intervals.sort()
        result = [intervals[0]]
        for i in range(1,len(intervals)):
            curr = intervals[i]
            last_join = result[-1]
            if curr[0]<=last_join[1]:
                last_join[1]=max(last_join[1],curr[1])
            else:
                result.append(curr)
        return result

File: 056. 合并区间.go
package main

import "sort"

func merge(intervals [][]int) [][]int {
	if len(intervals) == 0 {
		return [][]int{}
	}
	if len(intervals) == 1 {
		return intervals
	}

	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	res := [][]int{intervals[0]}
	for i := 1; i < len(intervals); i++ {
		if intervals[i][0] <= res[len(res)-1][1] {
			res[len(res)-1][1] = max(res[len(res)-1][1], intervals[i][1]) // !!!
			continue
		}
		res = append(res, intervals[i])
	}

	return res
}
